drivers/overlay/overlay.go

```diff
@@ -2329,17 +2346,25 @@ func (d *Driver) ApplyDiff(id, parent string, options graphdriver.ApplyDiffOpts)
 		return 0, err
 	}
 
-	logrus.Debugf("Applying tar in %s", applyDir)
-	// Overlay doesn't need the parent id to apply the diff
-	if err := untar(options.Diff, applyDir, &archive.TarOptions{
-		UIDMaps:           idMappings.UIDs(),
-		GIDMaps:           idMappings.GIDs(),
-		IgnoreChownErrors: d.options.ignoreChownErrors,
-		ForceMask:         d.options.forceMask,
-		WhiteoutFormat:    d.getWhiteoutFormat(),
-		InUserNS:          unshare.IsRootless(),
-	}); err != nil {
-		return 0, err
+	if options.LayerFilename != nil {
+		logrus.Debugf("Applying file in %s", applyDir)
+		err := cp(options.Diff, applyDir, *options.LayerFilename)
```

Comment:
`options.Diff` is an uncompressed stream (that the caller is interpreting as a tar). This _just doesn’t work_ for arbitrary blobs.